### PR TITLE
Simplified handling of shipkit repository

### PR DESCRIPTION
**Pull request by szczepiq**
- I have attempted to make GitHub repository setting optional in shipkit file (#285) but it is not possible. There are use cases where it is very useful to have have the repo name checked in to the source control. For example, it makes creation of pull requests from forks easier. Also, it makes some useful tasks like preview release notes useful when origin repository is forked.
 - Reverted back to the original idea @wwilk had
 - Tightened tests and documentation
 - Refactorings and rename jobs

### CODE REVIEW

**Comment by szczepiq** [comment]
Friendly ping :)

File: subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/git/tasks/IdentifyGitOriginRepoTask.java [hunk 1]
```diff
@@ -32,47 +32,41 @@ public IdentifyGitOriginRepoTask() {
 
     @TaskAction
     public void identifyGitOriginRepo() {
-        if (originRepo != null) {
-            LOG.lifecycle("  Using Git origin repository configured on the task: {}", originRepo);
+        if (repository != null) {
```
Comment (wwilk) [inline subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/git/tasks/IdentifyGitOriginRepoTask.java]: It would be good to have a short comment explaining when this situation can happen.

File: subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/notes/tasks/FetchContributors.java [hunk 2]
```diff
@@ -14,7 +14,7 @@
     private static final Logger LOG = Logging.getLogger(FetchContributors.class);
 
     public void fetchContributors(FetchGitHubContributorsTask task) {
-        LOG.lifecycle("  Fetching all contributors for project");
+        LOG.lifecycle("  Fetching all GitHub contributors of {}", task.getRepository());
```
Comment (wwilk) [inline subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/notes/tasks/FetchContributors.java]: 👍

File: subprojects/shipkit/src/test/groovy/org/shipkit/gradle/ShipkitGradlePluginIntegTest.groovy [hunk 3]
```diff
@@ -44,7 +44,6 @@ class ShipkitGradlePluginIntegTest extends GradleSpecification {
 :buildArchives
 :gitTag
 :identifyGitBranch
-:identifyGitOrigin
```
Comment (wwilk) [inline subprojects/shipkit/src/test/groovy/org/shipkit/gradle/ShipkitGradlePluginIntegTest.groovy]: So this task is not a part of performRelease now? Is it only used in initShipkitFileTask?

File: subprojects/shipkit/src/test/groovy/org/shipkit/internal/gradle/git/GitOriginPluginTest.groovy [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package org.shipkit.internal.gradle.git
+
+import testutil.PluginSpecification
+
+class GitOriginPluginTest extends PluginSpecification {
```
[the remark this answers is not in the document]
Comment (szczepiq) [reply subprojects/shipkit/src/test/groovy/org/shipkit/internal/gradle/git/GitOriginPluginTest.groovy]: > Missing test for scenario when plugin is applied but project is not the root.

I know I left a TODO for missing test :) I don't think it is blocking problem for the PR.

>Additionally in GitAuthPluginTest there was a test for provideAuthTo and it wasn't moved here

I must have removed it :) provideAuthTo method does not have any code worth testing. It's all declarative.

We should, however, add an integration test for shipkit init I think. That would be really cool. I wil add a TODO in the code.

File: subprojects/shipkit/src/test/groovy/org/shipkit/gradle/ShipkitGradlePluginIntegTest.groovy (hunk 3 above)
Comment (szczepiq) [reply subprojects/shipkit/src/test/groovy/org/shipkit/gradle/ShipkitGradlePluginIntegTest.groovy]: It is used in:
 - init scenarios
 - upgradeDownstream/versionUpgrade scenarios

We cannot really use it for other scenarios, e.g. to get rid of 'repository name' setting in shipkit file. I really tried :)

Example use case: if I fork mockito repository, clone it to my local, prepare a PR and want to preview the release notes, I need the release notes generation logic to use "mockito/mockito" repo, and not "szczepiq/mockito" repo. For now, we will keep the repository name setting in the shipkit file.

**Comment by szczepiq** [comment]
Thank you for feedback. Let me know what else I can do to ship this change. I'd rather not complicate this PR further if with small code optimizations / adding missing tests / etc. Happy to follow up with a PR to wrap things up.

File: subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/git/tasks/IdentifyGitOriginRepoTask.java (hunk 1 above)
Comment (szczepiq) [reply subprojects/shipkit/src/main/groovy/org/shipkit/internal/gradle/git/tasks/IdentifyGitOriginRepoTask.java]: Cool idea. Will do!

File: subprojects/shipkit/src/test/groovy/org/shipkit/internal/gradle/git/GitOriginPluginTest.groovy (hunk 4 above)
[the remark this answers is not in the document]
Comment (wwilk) [reply subprojects/shipkit/src/test/groovy/org/shipkit/internal/gradle/git/GitOriginPluginTest.groovy]: What about the order (checking if the action provided in argument is actually executed and in the correct moment after repository is already set) and checking if proper repository value is passed to the action?
I think this test here:
https://github.com/mockito/shipkit/pull/424/files#diff-630faa16ee3a0156b085a0e1cc134b61L8
does that. IMHO it's quite simple and does the job. But of course integration test would work too, the question is which is one is simpler/faster/easier to maintain. WDYT?

[the remark this answers is not in the document]
Comment (szczepiq) [reply subprojects/shipkit/src/test/groovy/org/shipkit/internal/gradle/git/GitOriginPluginTest.groovy]: Good feedback! Let's work it out. Here's the implementation suggested in the PR:

```java
        t.dependsOn(identifyTask);
        identifyTask.doLast(new Action<Task>() {
            public void execute(Task task) {
                originRepoName.execute(identifyTask.getRepository());
            }
        });
```

>What about the order (checking if the action provided in argument is actually executed and in the correct moment after repository is already set) and checking if proper repository value is passed to the action?

This seems to me like testing Gradle. I trust that Gradle works:
 - that "doLast" action will get triggered after task is executed
 - that "dependsOn" will ensure correct order

The single line of executing the action and setting the repo is not worth testing - this is just calling a method. It has zero complexity, no conditional logic, no parameter manipulation, no math, no String concat/regexp  - what kind of bug are we guarding against?

Let's role play some "bug scenarios":
 - let's say that during initial implementation, before submitting the PR, I forgot to write "dependsOn" and I did not test this properly e2e. Well if I forget something in the implementation, for sure I would forget it in the unit test, too. Only the integration test would have fleshed this out.
 - let's say that I will plant a bug in the future, I accidentally remove "dependsOn" and it is not detected in code review. Travis CI will fail with downstream tests (which are awesome BTW!!!) because without dependsOn, there is no repo information and stuff breaks.

In summary:
 - in general, we shouldn't write unit tests for a) Gradle functionality b) for code that has zero complexity and is fully "declarative".
 - we are already covered with downstream tests that act as integration tests

The less code, the better, and that includes test code :)

Given that the presence of a unit test is the only contention point and you seem to be happy with the other changes, I will ship this change and take the unit test discussion offline. I'm very happy to converge on the best testing strategy. Please convince me that I need a unit test for this.

Thank you for thoughtful feedback!

[the remark this answers is not in the document]
Comment (szczepiq) [reply subprojects/shipkit/src/test/groovy/org/shipkit/internal/gradle/git/GitOriginPluginTest.groovy]: @mockito/shipkit-developers, guys, can you weigh on this discussion about unit tests? It would be great to hear other opinions so that we can converge on cohesive testing strategy!

[the remark this answers is not in the document]
Comment (epeee) [reply subprojects/shipkit/src/test/groovy/org/shipkit/internal/gradle/git/GitOriginPluginTest.groovy]: first of all sorry for the delay, it looks like I somehow overlooked this one.
I do agree with @szczepiq when it comes to testing.  Let's have unit tests for code with high complexity and try to avoid tests for zero complexity/testing gradle functionality.
